### ISSUE TITLE
resolving @import

### ISSUE THREAD

**Issue by tcurdt** [stylus/stylus#2039]
While the docs here are great https://github.com/stylus/stylus/blob/dev/docs/import.md they left me wondering how to control the loading/resolving of imports. I am using stylus on Node:

      var stylus = Stylus("....");
      stylus.set('filename', file.path);
      stylus.render(function(err, result){
        ...
      });

Is this possible at all?

**Comment by Panya** [comment]
What do you mean by "loading/resolving"?

**Comment by tcurdt** [comment]
I want to tell Stylus where to find the imports - or even provide them.

Similar to https://mozilla.github.io/nunjucks/api.html#loader or https://github.com/less/less-plugin-npm-import/blob/master/lib/npm-file-manager.js

**Comment by Panya** [comment]
I see. Unfortunately Stylus' importer is not extensible in this way currently.

**Comment by tcurdt** [comment]
@Panya Urgh - that's unexpected

Stylus has one of the nicest APIs from the other alternatives out there. So how do we fix this oversight? :)

If you guide me a little I might be able to help.

**Comment by tcurdt** [comment]
Seems like this is where things need to change a little
https://github.com/stylus/stylus/blob/dev/lib/visitor/deps-resolver.js#L112

**Comment by Panya** [comment]
It will be a little bit harder :) We need to extract default logic for importing files from the evaluator. These methods are [importFile](https://github.com/stylus/stylus/blob/dev/lib/visitor/evaluator.js#L29-L100) (messy) and [visitImport](https://github.com/stylus/stylus/blob/dev/lib/visitor/evaluator.js#L854-L910).

**Comment by tcurdt** [comment]
@Panya well, that would have been too easy otherwise ;)

Maybe we start thinking about the API and then see how it could be implemented in those places.
Since I just looked at the various APIs of different libraries I have a pretty clear idea of what I would love to see:

    resolve = function(absPathFrom, includeArgument, cb) {
      var absPathTo = path.resolve(absPathFrom, includeArgument);
      var content = fs.readFileSync(absPathTo, 'utf-8');
      cb(null, content);
    }

The `absPathFrom` would be the absolute path of the file including and `includeArgument` the argument passed to the include statement.

Sounds sane?

**Comment by Panya** [comment]
Yep, but you also need to know that all internal Stylus' processes are currently synchronous, so that would limit the functionality to synchronous-only resolvers (otherwise we need to go deeper :)).

**Comment by tcurdt** [comment]
synchronous is fine - for me, for now :)
I would still use callback style for the API though

**Comment by SamHwang1990** [comment]
Hi~~~
Is there any ways to allow optinal @import file like less?
The mainly purpose of this demand is to let stylus to import a file which may be not existed without throwing error.

**Comment by tcurdt** [comment]
@SamHwang1990 that sounds like a very different issue. Better open an individual issue for it.

**Comment by tcurdt** [comment]
@Panya tag as enhancement request?

I will see if I can find some time to dig into this (fingers crossed) but as `less` and `sass` is working fine the urge is not that strong. Would love to support `stylus` though.

**Comment by SamHwang1990** [comment]
Sorry about my comment~~~
I have open a new issue about it.
Thank you! @tcurdt
